src/java.base/share/classes/java/math/MutableBigInteger.java

```diff
@@ -287,6 +290,9 @@ final int compare(MutableBigInteger b) {
      * would return, but doesn't change the value of {@code b}.
      */
     private int compareShifted(MutableBigInteger b, int ints) {
+        this.normalize();
```

Comment:
See [JDK-8334483](http://bugs.java.com/bugdatabase/view_bug?bug_id=JDK-8334483)

Reply:
@djelinski So, why the documentation does not specify that the method assumes the arguments are normalized? And why the method `subtract(MutableBigInteger)` does not call `normalize()` on arguments before calling `compare(MutableBigInteger)`? Also `subtract` assumes implicitly that the arguments are normalized?

Reply:
> why the documentation does not specify that the method assumes the arguments are normalized?

Well, the best answer I can offer is "because no one documented it yet". `MutableBigInteger` is an internal class, so most people never see its documentation.
> And why the method subtract(MutableBigInteger) does not call normalize() on arguments before calling compare(MutableBigInteger)? Does subtract assume implicitly that the arguments are normalized too?

Most likely. If you check the code you'll notice that most methods that change the `MutableBigInteger` value (like `subtract`) call `normalize` after performing the operation, so the values should be normalized most of the time.

Reply:
> Most likely. If you check the code you'll notice that most methods that change the `MutableBigInteger` value (like `subtract`) call `normalize` after performing the operation, so the values should be normalized most of the time.

@djelinski Unfortunately, this is not always true. In fact, without forcing normalization before comparing, the program leads to failure in thousands of cases, while forcing the normalization resolves all the failures...

Reply:
Which program are you referring to?

It is the responsibility of the caller of `subtract` to ensure that the arguments are normalized. This may or may not require calling `normalize`, depending on whether the argument is normalized or not.

Reply:
@djelinski I'm referring to the tests of square root for BigIntegers.

Reply:
ah, so before your changes, the arguments were always normalized, and now they are not?

Reply:
@djelinski Now I found the issue and resolved it normalizing the Zimmerman's blocks.